helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizerAccessor.java

```diff
@@ -0,0 +1,354 @@
+      return badRequest("Failed to compute partition assignment: " + e);
+    }
+  }
+
+  private InputFields readInput(String content)
+      throws InvalidParameterException, JsonProcessingException {
+
+    Map<String, Map<String, Object>> customFieldsMap;
+    InputFields inputMap = new InputFields();
+    customFieldsMap =
+        OBJECT_MAPPER.readValue(content, new TypeReference<HashMap<String, Map<String, Object>>>() {
+        });
+
+    // Content is given as a KV mapping.
+    for (Map.Entry<String, Map<String, Object>> entry : customFieldsMap.entrySet()) {
+      String key = entry.getKey();
+      switch (key) {
+        case INSTANCE_CHANGE:
+          for (Map.Entry<String, Object> instanceChange : entry.getValue().entrySet()) {
+            String instanceChangeKey = instanceChange.getKey();
+            if (instanceChangeKey.equals(INSTANCE_CHANGE_ADD_INSTANCES) && (instanceChange
+                .getValue() instanceof List)) {
+              inputMap.newInstances.addAll((List<String>) instanceChange.getValue());
+            } else if (instanceChangeKey.equals(INSTANCE_CHANGE_REMOVE_INSTANCES) && (instanceChange
+                .getValue() instanceof List)) {
+              inputMap.instancesToRemove.addAll((List<String>) instanceChange.getValue());
+            } else if (instanceChangeKey.equals(INSTANCE_CHANGE_SWAP_INSTANCES) && (instanceChange
+                .getValue() instanceof Map)) {
+              for (Map.Entry<String, String> swapPair : ((Map<String, String>) instanceChange
+                  .getValue()).entrySet()) {
+                inputMap.nodeSwap.put(swapPair.getKey(), swapPair.getValue());
+              }
+            } else {
+              throw new InvalidParameterException(
+                  "Unsupported command or invalid format for InstanceChange : "
+                      + instanceChangeKey);
+            }
+          }
+          break;
+        case OPTIONS:
+          for (Map.Entry<String, Object> option : entry.getValue().entrySet()) {
+            String optionKey = option.getKey();
+            if (optionKey.equals(OPTIONS_RESOURCE_FLT) && (option.getValue() instanceof List)) {
+              inputMap.resourceFilter.addAll((List<String>) option.getValue());
+            } else if (optionKey.equals(OPTIONS_INSTANCE_FLT) && (option.getValue() instanceof List)) {
+              inputMap.instanceFilter.addAll((List<String>) option.getValue());
+            } else {
+              throw new InvalidParameterException(
+                  "Unsupported command or invalid format for Options : " + option);
+            }
+          }
+          break;
+        default:
+          throw new InvalidParameterException(
+              "Unsupported command for partitionAssignment : " + key);
+      }
+    }
+    return inputMap;
+  }
+
+  private ClusterState readClusterStateAndValidateInput(String clusterId, InputFields inputFields)
```

Comment:
Have you explored the possibility of keeping a cluster data cache here? So we avoid read latency and we don't need to manually reconstruct a simplified rebalance logic here.

Let's add a TODO here at least.